pkg/ruler/rule_concurrency.go

```diff
@@ -0,0 +1,213 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ruler
+
+import (
+	"context"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/rules"
+	"golang.org/x/sync/semaphore"
+)
+
+type MultiTenantRuleConcurrencyController interface {
+	rules.RuleConcurrencyController
+}
+
+const IntervalToEvaluationThreshold = 50.00
+
+// DynamicSemaphore is a semaphore that can dynamically change its max concurrency.
+// It is necessary as the max concurrency is defined by the user limits which can be changed at runtime.
+type DynamicSemaphore struct {
+	logger         log.Logger
+	maxConcurrency func() int64
+
+	mtx      sync.Mutex
+	acquired int64
+}
+
+// NewDynamicSemaphore creates a new DynamicSemaphore
+func NewDynamicSemaphore(logger log.Logger, maxConcurrency func() int64) *DynamicSemaphore {
+	return &DynamicSemaphore{
+		logger:         logger,
+		maxConcurrency: maxConcurrency,
+		acquired:       0,
+	}
+}
+
+// TryAcquire tries to acquire a token from the semaphore.
+func (ds *DynamicSemaphore) TryAcquire() bool {
+	ds.mtx.Lock()
+	defer ds.mtx.Unlock()
+	maxTokens := ds.maxConcurrency()
```

Comment:
[nit] I would move this call before taking `ds.mtx.Lock()`, since it's not required to get the lock.

Reply:
Great shout, done in https://github.com/grafana/mimir/pull/8146/commits/004f3a11b2208bc5318a49460b1e10fc89d7e3bd